cmake/nccl.cmake

```diff
@@ -0,0 +1,30 @@
+if (NOT WITH_GPU)
+  return ()
+endif()
+
+set(NCCL_ROOT "/usr" CACHE PATH "CUDNN ROOT")
+find_path(NCCL_INCLUDE_DIR nccl.h PATHS
+        ${NCCL_ROOT} ${NCCL_ROOT}/include
+        $ENV{NCCL_ROOT} $ENV{NCCL_ROOT}/include ${CUDA_TOOLKIT_INCLUDE}
+        NO_DEFAULT_PATH)
+
+get_filename_component(__libpath_hist ${CUDA_CUDART_LIBRARY} PATH)
+
+set(TARGET_ARCH "x86_64")
+if(NOT ${CMAKE_SYSTEM_PROCESSOR})
+  set(TARGET_ARCH ${CMAKE_SYSTEM_PROCESSOR})
+endif()
+
```

Comment:
Just for curiosity, could we automatically download NCCL and put the header and library files into the `build/third_party` directory?  We don't have to do this in this PR, but could just file an issue recording this idea.

Reply:
`nccl` will be packaged to official CUDA docker image.

We can download `nccl` header in our Paddle source code. But to keep it simple, we just assume the developers are installed CUDA and related libraries well.

From @dzhwinter 's issue, https://gitlab.com/nvidia/cuda/issues/10#note_43614309

Reply:
No, we shouldn't.  This PR just moves NCCL from `build/third_party` to the docker built-in environment. Partly of this PR is cherry-picked from  https://github.com/PaddlePaddle/Paddle/pull/4818, which automate download NCCL and build it locally. In #4818, we have to build it early than any other module, so we need to add NCCL as a dependency, which is not correct comparing with DSO

Reply:
I think I understand your point -- some users might already have a special version of NCCL installed on their systems, so PaddlePaddle should try to use it.